lighthouse-core/audits/sized-images.js

```diff
@@ -0,0 +1,123 @@
+ */
+'use strict';
+
+const Audit = require('./audit.js');
+const i18n = require('./../lib/i18n/i18n.js');
+const URL = require('./../lib/url-shim.js');
+
+const UIStrings = {
+  /** Title of a Lighthouse audit that provides detail on whether all images had width and height attributes. This descriptive title is shown to users when every image has width and height attributes */
+  title: 'Image elements have `width` and `height` attributes',
+  /** Title of a Lighthouse audit that provides detail on whether all images had width and height attributes. This descriptive title is shown to users when one or more images does not have width and height attributes */
+  failureTitle: 'Image elements do not have `width` and `height` attributes',
+  /** Description of a Lighthouse audit that tells the user why they should include width and height attributes for all images. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. */
+  description: 'Always include width and height attributes on your image elements to reduce layout shifting and improve CLS. [Learn more](https://web.dev/optimize-cls/#images-without-dimensions)',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/**
+ * @fileoverview
+ * Audit that checks whether all images have width and height attributes.
+ */
+
+class SizedImages extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'sized-images',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['ImageElements'],
+    };
+  }
+
+  /**
+   * @param {string} attr
+   * @return {boolean}
+   */
+  static isValid(attr) {
+    // an img size attribute is valid for preventing CLS
+    // if it is a non-negative, non-zero integer
+    const NON_NEGATIVE_INT_REGEX = /^\d+$/;
+    const ZERO_REGEX = /^0+$/;
+    return NON_NEGATIVE_INT_REGEX.test(attr) && !ZERO_REGEX.test(attr);
+  }
+
+  /**
+   * @param {string} attrWidth
+   * @param {string} attrHeight
+   * @param {string} cssWidth
+   * @param {string} cssHeight
+   * @return {boolean}
+   */
+  static unsizedImage(attrWidth, attrHeight, cssWidth, cssHeight) {
+    // images are considered sized if they have defined & valid values
+    if (attrWidth && attrHeight) {
+      return !SizedImages.isValid(attrWidth) || !SizedImages.isValid(attrHeight);
+    }
+    if (attrWidth && cssHeight) {
+      return !SizedImages.isValid(attrWidth) || cssHeight === 'auto';
+    }
+    if (cssWidth && attrHeight) {
+      return cssWidth === 'auto' || !SizedImages.isValid(attrHeight);
```

Comment:
From the testing that I did, if the cssWidth/cssHeight weren't defined explicitly `getEffectiveSizingRule` would return `undefined`, so I normalized the default return to be `''` so that the value of cssWidth/cssHeight stays consistent with its interface string value, & because of short-circuit `&&` I didn't write much further validation testing with`isValidCss` 

does this sound like a valid approach?

Reply:
Yeah that behavior in the gatherer sounds great 👍  (though probably worth adding a note to the gatherer typedef about how it does not use the default CSS value)

It just means we need to do a little more work here in the audit. `cssWidth === 'auto'` would only catch explicitly set `width: auto` from what you've just described to me and we would want to catch *all* the cases where we don't have a css width value (`undefined|''|'auto'`)

Reply:
To clarify, this is my plan based on that guidance

1. change the typedef to be something like `cssWidth: string | undefined`
2. remove my default normalization to `''`  
3. explicitly check for `undefined | '' | 'auto'` in `isValidCss` 
4. I am still confused about whether it is redundant to explicitly check for `undefined | ''` since they are falsy / would short-circuit the `&&` logic. Am I misunderstanding something?

Let me know what you think!

Reply:
Oooooooooh sorry I was totally missing what you were saying @lemcardenas 🤦 yeah the short circuiting definitely takes care of the `''` part. I'd be +1 to an explicit check in `isValidCss` :)

do invalid CSS values also return `undefined`?

Reply:
thanks for mentioning that case because I just checked and invalid CSS values are not returned as `undefined`/`''`

so it looks like I might have to regex for valid values... it would be almost the same logic as the `isValidAttr` except for an optional suffix for units & a separate check for percentages

when looking through the possible unit suffixes, should I limit them to a certain subset? There are many possibilities I had never heard of according to [this](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units) although the documentation [here](https://www.w3.org/TR/CSS2/syndata.html#value-def-length) shows a much more limited subset.

Reply:
I would err on the side of being the most permissive to avoid false positives. It wouldn't be out of the question IMO, to stick to your original logic only failing `'auto'`|`undefined`|`''`. My question was mostly for us documenting properly that it contains the raw, unvalidated css property value :) 

Something permissive like `[0-9.]+(px|em|vw|%|...)?` with all the unit options from [MDN](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units) could work too (that other CSS spec link looks like it's for an old version, many of the missing ones there are quite common).

Reply:
sounds good! to simplify things, I am going to go with the original logic of only failing `'auto'|undefined|''`. 

I might tryout a regex later once we have a tool to run lighthouse on a dataset to see if there are noticeable differences in false positives with each approach

Reply:
Great idea 👍